src/DataStructures/ModalVector.hpp

```diff
@@ -92,302 +136,221 @@ class ModalVector {
+                const blaze::DenseVector<VT, VF>& rhs) noexcept {
+  return not(lhs == rhs);
+}
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) operator/(
-      const ModalVector& lhs, const double& rhs) noexcept {
-    return lhs.data_ / rhs;
-  }
-  template <typename VT, bool VF>
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) operator/(
-      const ModalVector& lhs, const blaze::Vector<VT, VF>& rhs) noexcept {
-    return lhs.data_ / ~rhs;
-  }
-  // FIXME PK: Can we avoid the instantiation of new DV object in this function
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) operator/(
-      const ModalVector& lhs, const DataVector& rhs) noexcept {
-    const blaze::DynamicVector<double> _rhs_local(rhs.size(), rhs.data());
-    return lhs / _rhs_local;
-  }
+template <typename VT, bool VF>
+bool operator==(const blaze::DenseVector<VT, VF>& lhs,
+                const ModalVector& rhs) noexcept {
+  return ModalVector(lhs) == rhs;
+}
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) min(
-      const ModalVector& t) noexcept {
-    return min(t.data_);
-  }
+template <typename VT, bool VF>
+bool operator!=(const blaze::DenseVector<VT, VF>& lhs,
+                const ModalVector& rhs) noexcept {
+  return not(lhs == rhs);
+}
+/// \endcond
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) max(
-      const ModalVector& t) noexcept {
-    return max(t.data_);
-  }
+// Specialize the Blaze type traits to correctly handle ModalVector
+namespace blaze {
+template <>
+struct IsVector<ModalVector> : std::true_type {};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) abs(
-      const ModalVector& t) noexcept {
-    return abs(t.data_);
-  }
+template <>
+struct TransposeFlag<ModalVector> : BoolConstant<
+                ModalVector::transpose_flag> {};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) fabs(
-      const ModalVector& t) noexcept {
-    return abs(t.data_);
-  }
-  // @}
+template <>
+struct AddTrait<ModalVector, ModalVector> {
+  using Type = ModalVector;
+};
 
+template <>
+struct AddTrait<ModalVector, double> {
+  using Type = ModalVector;
+};
 
- private:
-  /// \cond HIDDEN_SYMBOLS
-  size_t size_ = 0;
-  InternalStorage_t owned_data_;
-  InternalModalVector_t data_;
-  bool owning_{true};
-  /// \endcond
+template <>
+struct AddTrait<double, ModalVector> {
+  using Type = ModalVector;
 };
 
-/// Output operator for ModalVector
-std::ostream& operator<<(std::ostream& os, const ModalVector& d);
+template <>
+struct SubTrait<ModalVector, ModalVector> {
+  using Type = ModalVector;
+};
 
-/// Equivalence operator for ModalVector
-bool operator==(const ModalVector& lhs, const ModalVector& rhs);
+template <>
+struct SubTrait<ModalVector, double> {
+  using Type = ModalVector;
+};
 
-/// Inequivalence operator for ModalVector
-bool operator!=(const ModalVector& lhs, const ModalVector& rhs);
+template <>
+struct SubTrait<double, ModalVector> {
+  using Type = ModalVector;
+};
+
+template <>
+struct MultTrait<ModalVector, ModalVector> {
```

Comment:
does this mean the result of multiplying two ModalVectors is a ModalVector?  If so, I don't like it